File: Programmers/Python/Lv.3/2 x n 타일링.py
# Dynamic Programming
# 2 * n -> 2 * (n-1)에서 세로타일을 하나 붙이는 경우 d[n-1]
# 2 * (n-2) 에서 가로타일 두개를 붙이는 경우 d[n-2]
# 점화식 -> d[n] = d[n-1] + d[n-2]

def solution(n):
    answer = 0
    dp = []
    dp.append(1)
    dp.append(2)
    
    for i in range(2, n+1) :
        dp.append((dp[i-1] + dp[i-2])%1000000007) 
    answer = dp[n-1]
    
    return answer

File: Programmers/Javascript/Lv.2/2xn 타일링.js
function solution(n) {
  const dp = Array.from({ length: n + 1 }, () => 0);
  dp[1] = 1;
  dp[2] = 2;
  for (let i = 3; i <= n; i++) {
    dp[i] = (dp[i - 2] + dp[i - 1]) % 1000000007;
  }
  return dp[n];
}

// push 사용하면 시간초과, 필요한 메모리 만큼만 할당
